### ISSUE TITLE
Have some way of detecting whether an expansion is top level or not

### ISSUE THREAD

**Issue by DavidDudson** [scalameta/paradise#159]
After my investigation into https://github.com/scalameta/paradise/issues/158...

## Problem

The issues that are produced due to top level macro expansion very much targeted at macro authors, not macro users.

The user of the macro probably will not understand the error shown here: 
https://github.com/scalameta/paradise/blob/master/plugin/src/main/scala/org/scalameta/paradise/typechecker/Errors.scala#L92

As a macro author would like to just have my user see a message similar to:

`AbortException: Macro foo cannot annotate a top level object. the annotee must be inside an object`

As the macro author will definitely know whether or not the macro will expand at the top level or not. So it would nice if I could give this information to the user easily.

It looks as if all of the errors are emitted during a validation phase post expansion, *not* during expansion.
Also, there's no way of telling whether or not the annotee is top level. `tree.parent` will return None. This is of course because the parent hasn't had its macros expanded yet.

## Potential solution

Introduce a new field into Tree `isTopLevel: Option[Boolean]`

When creating trees, or when unknown, this is `None`

When the tree is being passed into a macro annotation, the field is to populate with `true` or `false` using the same logic as the current validate method.

That way the macro author can add custom logic to macro annotations at the top level

**Comment by olafurpg** [comment]
Did you try `.parent`? You can traverse up to the top to see if a node is top-level or not.

**Comment by DavidDudson** [comment]
Yes. because `toMTree` is called on whatever the annotee is, the annotee always has no parent.

This is because Macro annotations work `in => out`

Perhaps we could call `toMtree` on the `gTree.parent` of the annotee. Then populate the annotees parent field. This tree would include the current representation of the annotee as well as anything else.

It would basically be the current view of the parent (Because expansion could affect them later.)

**Comment by olafurpg** [comment]
I don't think it's a good idea to add another method to Tree since this is a macro specific issue. I feel like macro annotations should get some sort of compiler context to for example emit compiler warnings. Maybe we can put this boolean bit in the context?

**Comment by xeno-by** [comment]
@DavidDudson @olafurpg In my dissertation (https://dl.dropboxusercontent.com/u/10497693/dissertation.pdf), I suggest that apart from `Dialect` and `Mirror`, we also introduce `Expansion` that contains methods like that. More precisely, see section 9.4. Meta APIs. Wdyt?

**Comment by DavidDudson** [comment]
@xeno-by sounds good
